apps/api/src/http/internal/rest/webhooks/stripe/invoice.rs

```diff
@@ -0,0 +1,314 @@
+			return Ok(None);
+		}
+
+		let product = products.into_iter().next().unwrap();
+
+		// This invoice is for one of our subscription products.
+
+		// Retrieve subscription
+		let stripe_sub = stripe::Subscription::retrieve(&global.stripe_client, &subscription.id(), &[])
+			.await
+			.map_err(|e| {
+				tracing::error!(error = %e, "failed to retrieve subscription");
+				TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+			})?;
+
+		let user_id = stripe_sub
+			.metadata
+			.get("USER_ID")
+			.and_then(|i| UserId::from_str(i).ok())
+			.ok_or(TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR))?;
+
+		let sub_id = SubscriptionId {
+			user_id,
+			product_id: product.id,
+		};
+
+		let start = chrono::DateTime::from_timestamp(stripe_sub.current_period_start, 0)
+			.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?;
+		// when the subscription is in trial, the current period end is the trial end
+		let end = chrono::DateTime::from_timestamp(stripe_sub.current_period_end, 0)
+			.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?;
+
+		let provider_id = ProviderSubscriptionId::from(stripe_sub.id);
+
+		tx.insert_one(
+			SubscriptionPeriod {
+				id: SubscriptionPeriodId::new(),
+				subscription_id: sub_id.clone(),
+				provider_id: Some(provider_id),
+				start,
+				end,
+				is_trial: stripe_sub.trial_end.is_some(),
+				created_by: SubscriptionPeriodCreatedBy::Invoice {
+					invoice_id: invoice.id.clone().into(),
+					cancel_at_period_end: stripe_sub.cancel_at_period_end,
+				},
+				updated_at: chrono::Utc::now(),
+				search_updated_at: None,
+			},
+			None,
+		)
+		.await?;
+
+		updated(global, tx, invoice, HashMap::new()).await?;
+
+		return Ok(Some(sub_id));
+	}
+
+	updated(global, tx, invoice, HashMap::new()).await?;
+
+	Ok(None)
+}
+
+/// Only sent for draft invoices.
+/// Deletes the invoice object.
+pub async fn deleted(
+	_global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	invoice: stripe::Invoice,
+) -> TransactionResult<(), ApiError> {
+	let id: InvoiceId = invoice.id.into();
+
+	tx.find_one_and_delete(
+		filter::filter! {
+			Invoice {
+				#[query(rename = "_id")]
+				id,
+			}
+		},
+		None,
+	)
+	.await?;
+
+	Ok(())
+}
+
+/// Marks the associated invoice as failed.
+/// Shows the user an error message.
+/// Should prompt the user to collect new payment information and update the
+/// subscription's default payment method afterwards.
+pub async fn payment_failed(
+	_global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	invoice: stripe::Invoice,
+) -> TransactionResult<(), ApiError> {
+	// TODO: Show the user an error message.
+	// TODO: Collect new payment information and update the subscriptions default
+	// payment method.
+
+	let id: InvoiceId = invoice.id.into();
+
+	tx.update_one(
+		filter::filter! {
+			Invoice {
+				#[query(rename = "_id")]
+				id,
+			}
+		},
+		update::update! {
+			#[query(set)]
+			Invoice {
+				failed: true,
```

Comment:
is there an invoice status here??
Perhaps we should also have a failed reason attached to the invoice?

Reply:
Yes, but the invoice status doesn't represent if one payment failed for it